### DISCUSSION TITLE
VF2 edge filtering

### DISCUSSION THREAD

**Discussion by aanastasiou** [networkx/networkx#6164] [Q&A] [answered]
The VF2 implementation in networkx returns a dictionary that maps nodes across the "pattern" and "source" graphs. It is also possible to provide an edge "criterion" to the constructor too which is great when a given pattern also has constraints on the edges.

But, in the MultiDiGraph case, how are the "matching edges" represented in that node mapping?

(One hack I can think of is to create an intermediate graph that contains ONLY the edges that would pass the criterion which then means that extracting a subgraph of a given result would only contain the relevant edges....but I don't know how well would this scale for large(ish) graphs...)

**Comment by dschult** [comment, marked as answer]
The edge_match feature allows you to provide a function which determines whether two edges are considered isomorphic. 

```
        edge_match: callable
            A function that returns True iff the edge attribute dictionary for
            the pair of nodes (u1, v1) in G1 and (u2, v2) in G2 should be
            considered equal during the isomorphism test. The function will be
            called like::
               edge_match(G1[u1][v1], G2[u2][v2])
            That is, the function will receive the edge attribute dictionaries
            of the edges under consideration. If None, then no attributes are
            considered when testing for an isomorphism.
```
Because the inputs to the function are the dictionaries `G[u][v]`, for MultiGraphs this is the keydict. That is, it is the dict keyed by edge key to the edgedata dictionary which holds edge attributes. The "edges" in this case represent pairs of nodes and the edge data passed in contains all the information about the multiedges between those two nodes. 

So, you can write your function to make it handle multiedges however you like. Simple examples are to ensure that both node-pairs have the same number of edges between them.  But if the edges have an edge attribute, you could make sure that the distribution of edge attributes across the multiedges match for that pair of nodes.

**Reply by aanastasiou** [reply]
I think that part is clear. 

What I am trying to say is this:

1. I am constructing a MultiDiGraphMatcher and pass `node_match` and `edge_match` functions.
2. I call `subgraph_isomorphisms_iter()` and receive a set of (correct) results
3. The result is a dictionary that maps nodes in the search graph to nodes in the pattern graph. For example:
    `{"n2":10, "n4":11, "n5":12}`

The reason I am writing "(corect) results" above is because this node mapping depends on the `edge_match` function. The results can change if the edge criteria change.

Now, we want to extract the `{"n2", "n4", "n5"}` subgraph from the above result. If we just extract this subgraph without further edge checking, we are taking along all of the edges that possibly have been defined between any two nodes from that result set.

And hence the question: Since we are already passing a function to the matcher which it uses to only include edges that pass the function's criteria, how can we propagate these edges to the result set too?

Otherwise, from what I can gather from your answer, I should get a result, iterate all pairs of nodes, apply the function to filter the edges and reconstruct the subgraph that the result points to. Is that correct?

**Reply by dschult** [reply]
Yes, the current returned isomorphism gives the mapping of nodes, so you would need to check the edges using the filter to reconstruct the subgraph that the result points to.

It sounds like this kind of feature might be one for the discussion #4453 about how to improve the tools for motif exploration.  What kinds of tools would be helpful?

**Reply by aanastasiou** [reply]
@dschult thank you for confirming. I will proceed with "manual" edge checking for the moment.

Re: 4453, just to clarify that this question (6164) was motivated from a different use case than the comment on 4453...we can talk about both of course.
